[1.18] Add getCodec method in ForgeRegistry

Frinn38 opened the pull request:
1.18 version of #7668 (1.16) and #8117 (1.17)
Just renamed the test mod from `CodecRegistryTest` to `RegistryCodecTest` since it makes more sense.
The rest is exactly the same as the 1.17 version.

Frinn38 committed:
Update registry codec PR to 1.18

alcatrazEscapee commented:
~~It's worth mentioning in 1.18, vanilla separated out the functionality of by-name and by-id encoding, rather than relying on the `ops.compressMaps()` flag. So they now have `Registry.byNameCodec()` and `Registry.byIdCodec()` for explicit purposes.~~

Edit: I was incorrect, as per below `Registry.byNameCodec` exists but it still uses the `ops.compressMaps()` flag.

Frinn38 commented:
I can't find `Registry.byIdCodec()` in my 1.18.1 workspace.
All I see is the `Registry.byNameCodec()` method, which make 2 codecs (compressed and not) and pass them to `ExtraCodecs.orCompressed()` which return a new codec that will use one of the two passed codecs depending on `ops.compressMaps()`
So basically the same as this PR except it create 3 `Codec` instance each time the method is called instead of only 1 when the registry is created.

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -871,6 +883,45 @@ public Snapshot makeSnapshot()
+        public <T> DataResult<T> encode(V input, DynamicOps<T> ops, T prefix)
+        {
+            ResourceLocation key = getKey(input);
+            if(key == null)
```
Formatting: space before opening parenthesis.
```suggestion
            if (key == null)
```

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -871,6 +883,45 @@ public Snapshot makeSnapshot()
+                return ops.getNumberValue(input).flatMap(n ->
+                {
+                    int id = n.intValue();
+                    if(ids.get(id) == null)
```
Formatting: space before opening parenthesis.
```suggestion
                    if (ids.get(id) == null)
```

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -871,6 +883,45 @@ public Snapshot makeSnapshot()
+        @Override
+        public <T> DataResult<Pair<V, T>> decode(DynamicOps<T> ops, T input)
+        {
+            if(ops.compressMaps())
```
Formatting: space before opening parenthesis.
```suggestion
            if (ops.compressMaps())
```

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -871,6 +883,45 @@ public Snapshot makeSnapshot()
+                    int id = n.intValue();
+                    if(ids.get(id) == null)
+                    {
+                        return DataResult.error("Unknown registry id " + n);
```
Could this and the other errors include the name of the registry as well, to align with the error messages from `Registry#byNameCodec`?

Frinn38 replied:
Done

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/ForgeRegistry.java:
```diff
@@ -871,6 +883,45 @@ public Snapshot makeSnapshot()
+            {
+                return ResourceLocation.CODEC.decode(ops, input).flatMap(keyValuePair -> !ForgeRegistry.this.containsKey(keyValuePair.getFirst()) ?
+                        DataResult.error("Unknown registry key: " + keyValuePair.getFirst()) :
+                        DataResult.success(keyValuePair.mapFirst(ForgeRegistry.this::getValue)));
```
I suggest moving the ternary operators a bit to make it more readable:
```suggestion
                return ResourceLocation.CODEC.decode(ops, input).flatMap(keyValuePair -> !ForgeRegistry.this.containsKey(keyValuePair.getFirst())
                        ? DataResult.error("Unknown registry key: " + keyValuePair.getFirst())
                        : DataResult.success(keyValuePair.mapFirst(ForgeRegistry.this::getValue)));
```

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/misc/RegistryCodecTest.java:
```diff
@@ -0,0 +1,89 @@
+     *         "block": "block_registry_name",
+     *         "item": "item_registry_name"
+     *     }
+     */
```
```suggestion
    /**
     * This Codec can serialize and deserialize a {@code Pair<Item, Block>}.
     * The resulting JSON (or NBT equivalent) will have this structure:
     * <pre>{@code
     * {
     *     "block": "block_registry_name",
     *     "item": "item_registry_name"
     * }
     * }</pre>
     */
```

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/misc/RegistryCodecTest.java:
```diff
@@ -0,0 +1,89 @@
+
+        //Serialize the Pair to NBT, and log an info in case of success or a warning in case of error
+        DataResult<Tag> result2 = CODEC.encodeStart(NbtOps.INSTANCE, pair);
+        result2.resultOrPartial(LOGGER::warn).ifPresent(tag -> LOGGER.info("Successfully encoded a Pair<Block, Item> to a nbt tag: " + tag));
```
Use logger argument substitution instead of manual string concatenation.
```suggestion
        result2.resultOrPartial(LOGGER::warn).ifPresent(tag -> LOGGER.info("Successfully encoded a Pair<Block, Item> to a nbt tag: {}", tag));
```

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/misc/RegistryCodecTest.java:
```diff
@@ -0,0 +1,89 @@
+        //Serialize the Pair to JSON using the COMPRESSED JsonOps, this will use the int registry id instead of the ResourceLocation one,
+        // this is not recommended because int IDs can change, so you should not rely on them
+        DataResult<JsonElement> result3 = CODEC.encodeStart(JsonOps.COMPRESSED, pair);
+        result3.resultOrPartial(LOGGER::warn).ifPresent(compressedJson -> LOGGER.info("Successfully encoded a Pair<Block, Item> to a compressed json: " + compressedJson));
```
Use logger argument substitution instead of manual string concatenation.
```suggestion
        result3.resultOrPartial(LOGGER::warn).ifPresent(compressedJson -> LOGGER.info("Successfully encoded a Pair<Block, Item> to a compressed json: {}", compressedJson));
```

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/misc/RegistryCodecTest.java:
```diff
@@ -0,0 +1,89 @@
+        FMLJavaModLoadingContext.get().getModEventBus().addListener(this::commonSetup);
+    }
+
+    public void commonSetup(final FMLCommonSetupEvent event)
+    {
```
Could you please add a test case for decoding using the compressed JSON ops? Here's one I made to test it myself:
```java
JsonArray jsonCompressed = new JsonArray();
jsonCompressed.add(((ForgeRegistry<Block>) ForgeRegistries.BLOCKS).getID(Blocks.DIAMOND_BLOCK));
jsonCompressed.add(((ForgeRegistry<Item>) ForgeRegistries.ITEMS).getID(Items.DIAMOND_PICKAXE));

DataResult<Pair<Pair<Block, Item>, JsonElement>> resultCompressed = CODEC.decode(JsonOps.COMPRESSED, jsonCompressed);
resultCompressed.resultOrPartial(LOGGER::warn).ifPresent(pair -> LOGGER.info("Decoded {} and {}", pair.getFirst().getFirst().getRegistryName(), pair.getFirst().getSecond().getRegistryName()));
```
You could also pipe in as the input the output of the compressed JSON ops in the last test case instead.

Frinn38 replied:
Added your test case

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/misc/RegistryCodecTest.java:
```diff
@@ -0,0 +1,89 @@
+import org.apache.logging.log4j.Logger;
+
+@Mod("registry_codec_test")
+public class RegistryCodecTest
```
Please add a javadoc comment to the test mod explaining what this test mod adds, what it does, and what it tests for.

Frinn38 replied:
Done

sciwhiz12 reviewed (changes requested):
The code looks solid to me, just mostly needs some formatting changes.

---

Please update the license headers added or modified by this PR to reflect the current year.

Frinn38 committed:
Apply suggestions from code review

Co-authored-by: sciwhiz12 <sciwhiz12@sciwhiz12.tk>

Frinn38 committed:
Merge branch '1.18.x' into 1.18.x-registry-codec

Frinn38 committed:
Apply requested changes

Update license header date to 2022.
Add test case to decode compressed json.
Add javadoc comment to explain what the test mod is doing.
Add the name of the registry in the codec error.

Frinn38 commented:
Updated license date to 2022 and implemented the requested changes.

sciwhiz12 reviewed (changes requested):
The current changes look good to me; only need to update the license header year for `src/main/java/net/minecraftforge/registries/ForgeRegistry.java` and `src/main/java/net/minecraftforge/registries/IForgeRegistry.java` to `2022`.

Frinn38 commented:
woops sorry the page didn't refresh when you replied, you are really fast ^^

Frinn38 committed:
Update license date

Update IForgeRegistry and ForgeRegistry license date to 2022
